bootstore/src/schemes/v0/peer.rs

```diff
@@ -850,11 +893,82 @@ mod tests {
         learner_jh.await.unwrap();
         handle0.load_rack_secret().await.unwrap_err();
 
-        // TODO: Once we have persistence, we can bring an old node back from the dead
-        // and reload the rack secret.
+        // Reload an node from persistent state and successfully reload the
+        // rack secret.
+        let (mut node1, handle1) = Node::new(config[1].clone(), &log).await;
+        let jh1 = tokio::spawn(async move {
+            node1.run().await;
+        });
+        let _ = handle1.load_peer_addresses(addrs.clone()).await;
+        handle0.load_rack_secret().await.unwrap();
+
+        // Add a second learner and ensure that the generation gets bumped for
+        // either node0 or node1
+        let peer0_gen =
+            handle0.get_status().await.unwrap().fsm_ledger_generation;
+        let peer1_gen =
+            handle1.get_status().await.unwrap().fsm_ledger_generation;
+        let learner_config = learner_config(&tempdir, 2);
+        let (mut learner, learner_handle) =
+            Node::new(learner_config.clone(), &log).await;
+        let learner_jh = tokio::spawn(async move {
+            learner.run().await;
+        });
+
+        // Inform the learner, node0, and node1 about all addresses including
+        // the learner. This simulates DDM discovery
+        addrs.insert(learner_config.addr.clone());
+        let _ = learner_handle.load_peer_addresses(addrs.clone()).await;
+        let _ = handle0.load_peer_addresses(addrs.clone()).await;
+        let _ = handle1.load_peer_addresses(addrs.clone()).await;
 
-        // Shutdown node0
+        // Tell the learner to go ahead and learn its share.
+        learner_handle.init_learner().await.unwrap();
+
+        // Get the new generation numbers
+        let peer0_gen_new =
+            handle0.get_status().await.unwrap().fsm_ledger_generation;
+        let peer1_gen_new =
+            handle1.get_status().await.unwrap().fsm_ledger_generation;
+
+        // Ensure only one of the peers generation numbers gets bumped
+        assert!(
+            (peer0_gen_new == peer0_gen && peer1_gen_new == peer1_gen + 1)
+                || (peer0_gen_new == peer0_gen + 1
+                    && peer1_gen_new == peer1_gen)
+        );
+
+        // Wipe the learner ledger, restart the learner and instruct it to
+        // relearn its share, and ensure that the neither generation number gets
+        // bumped because persistence doesn't occur.
```

Comment:
Why doesn't persistence occur?  I believe it's because the learner will ask the same peer for a share, and that peer already knows that it distributed that share to the learner?

Reply:
It's both. Due to the ordering (via internal BTreeSet) the learner will always ask the same peer first. That peer will also have handed out a share already to that peer. It marks that share as belonging to that peer in [distributed_shares](https://github.com/oxidecomputer/omicron/blob/main/bootstore/src/schemes/v0/fsm.rs#L57) and persisted that fact. The code ensures the same share gets handed out, which doesn't require persistence. We don't want to persist to limit any unnecessary write cycles on the M.2s. 

As a side note, this mechanism is actually intended to solve another problem: If the learner keeps disconnecting /timing out after asking a sled for a share but before receiving it. It will perpetually rotate among sleds asking for a share. By restricting each sled to hand out the same share to a given learner we limit each sled from handing out all there shares by accident and exhausting our pool. So at most we hand out `N-1` extra shares where `N` is the number of initial members in the trust quorum cluster.